kernel/init.c

```diff
@@ -252,20 +252,26 @@ static void z_sys_init_run_level(enum init_level level)
 		const struct device *dev = entry->dev;
 
 		if (dev != NULL) {
-			int rc = entry->init_fn.dev(dev);
-			/* Mark device initialized.  If initialization
-			 * failed, record the error condition.
-			 */
-			if (rc != 0) {
-				if (rc < 0) {
-					rc = -rc;
+			int rc = 0;
+
+			if (entry->init_fn.dev != NULL) {
+				rc = entry->init_fn.dev(dev);
+				/* Mark device initialized. If initialization
+				 * failed, record the error condition.
+				 */
+				if (rc != 0) {
+					if (rc < 0) {
+						rc = -rc;
+					}
+					if (rc > UINT8_MAX) {
+						rc = UINT8_MAX;
+					}
+					dev->state->init_res = rc;
 				}
-				if (rc > UINT8_MAX) {
-					rc = UINT8_MAX;
-				}
-				dev->state->init_res = rc;
 			}
+
 			dev->state->initialized = true;
```

Comment:
Would it be possible to ignore the initialisation level completely for devices without an initialisation function - and just statically mark the as "initialised"?

Reply:
Hmm, how would you do that?

Reply:
> @gmarull Hmm, how would you do that?

Probably in Z_DEVICE_DEFINE we can set state in Z_DEVICE_BASE_DEFINE and we can drop creating init entry with Z_DEVICE_INIT_ENTRY_DEFINE.

However I'm not sure if we can integrate it in existing Z_DEVICE_DEFINE macro without defining new one for the devices which doesn't require initialization.

Reply:
The problem I see is that we cannot evaluate at compile time if `init_fn` is `NULL` and make a decision based on that. We should probably introduce a new macro like `DEVICE_DEFINE_NOINIT`, that skips the creation of a device init entry and sets initialized = true.

Reply:
It's probably not worth it to do this at link time after the first stage link, is it?